PyDSS/controllers.py

```diff
@@ -0,0 +1,333 @@
+        Parameters
+        ----------
+        controller_model: ControllerBaseModel
+        element_names: list | None
+            List of element names (str). If None, use all elements for the given class.
+
+        """
+        self._controller_model = controller_model
+        if element_names is None:
+            self._element_names = list_element_names_by_class(self.get_element_class())
+        else:
+            self._element_names = element_names
+
+    @property
+    def controller_model(self):
+        return self._controller_model
+
+    @property
+    def element_names(self):
+        return self._element_names
+
+    def get_controller_class(self):
+        return type(self._controller_model).get_controller_class()
+
+    def get_element_class(self):
+        return type(self._controller_model).get_element_class()
+
+
+class ControllerManager:
+    """Provides ability to run control algorithms on circuit elements."""
+
+    def __init__(self, controllers: dict, solver: solver_base, max_control_iterations: int, error_tolerance: float):
+        self._controllers = controllers
+        self._max_control_iterations = max_control_iterations
+        self._error_tolerance = error_tolerance
+        self._solver = solver
+        self._stats = TimerStatsCollector()
+
+    @classmethod
+    def create(cls, controllers: list, settings: ProjectModel):
+        """Create controllers. The circuit must be loaded in OpenDSS."""
+        solver = get_solver_from_simulation_type(settings)
+        buses = OpenDSS.CreateBusObjects()
+        elements, elements_by_class = OpenDSS.CreateDssObjects(buses)
+        controllers_by_class = defaultdict(dict)
+        for circuit_element_controller in controllers:
+            controller_class = circuit_element_controller.get_controller_class()
+            element_class = circuit_element_controller.get_element_class()
+            for name in circuit_element_controller.element_names:
+                element = elements.get(name)
+                if element is None:
+                    raise InvalidConfiguration(f"{name} is not in the circuit")
+                controller = controller_class(
+                    element,
+                    circuit_element_controller.controller_model.dict(by_alias=True),
+                    dss,
+                    elements,
+                    solver,
+                )
+                controllers_by_class[element_class]["Controller." + name] = controller
+
+        return cls(
+            controllers_by_class,
+            solver,
+            settings.max_control_iterations,
+            settings.error_tolerance,
+        )
+
+    def run_controls(self):
+        """Run all controls.
+
+        Returns
+        -------
+        bool
+            True if all controllers converged, otherwise False.
+
+        """
+        has_converged = True
+        with Timer(self._stats, "RunControls"):
+            for priority in range(CONTROLLER_PRIORITIES):
```

Comment:
There is some duplication with the code in dssInstance, but I don't see a good way around it without performing a substantial rewrite of the existing code.